mlir/lib/Dialect/Affine/Transforms/AffineLoopInterchange.cpp

```diff
@@ -0,0 +1,964 @@
+
+  SmallVector<unsigned, 4>
+  getBestPermutation(SmallVector<AffineForOp, 4> &AffineLoopNest,
+                     DenseMap<Value, unsigned> &indVarIdMap);
+
+  void clear(SmallVector<AffineForOp, 4> &loopNest,
+             DenseMap<Value, unsigned> &indVarIdMap);
+
+  constexpr static unsigned cache_line_size = 64;
+
+  /// For each memref access function Ax+B, this is
+  /// the collection of all A's indexed by their
+  /// respective affine.load/affine.store operations.
+  DenseMap<Operation *, SmallVector<SmallVector<int64_t, 4>, 4>>
+      loopAccessMatrices;
+
+  /// For a memref access function Ax+B, this represents B.
+  DenseMap<Operation *, SmallVector<int64_t, 4>> constVector;
+
+  /// Element size (in bytes) of a load/store operation.
+  DenseMap<Operation *, unsigned> elementSize;
+
+  /// Iteration count for each affine.for in the loop nest.
+  SmallVector<unsigned, 4> loopIterationCounts;
+
+  /// Loop Carried Dependency vector. A 'true' at index 'i'
+  /// means loop at depth i carries a dependence.
+  SmallVector<bool, 4> loopCarriedDependenceVector;
+
+  /// Number of cache lines accessed (locality) by each affine.for
+  /// if it is considered the innermost loop.
+  DenseMap<const AffineForOp *, long double> numberOfCacheLinesAccessed;
+
+  /// affine.load/affine.store ops exibiting group-spatial/group-temporal
+  /// reuse.
+  SmallVector<llvm::SmallSet<Operation *, 8>, 8> referenceGroups;
+};
+} // namespace
+
+/// Returns True if any affine.if op found in the loop nest rooted
+/// at `forOp`
+static bool hasAffineIfStatement(AffineForOp &forOp) {
+  auto walkResult =
+      forOp.walk([&](AffineIfOp op) { return WalkResult::interrupt(); });
+  return walkResult.wasInterrupted();
+}
+
+/// Checks if this loop nest has a rectangular shaped iteration
+/// space.
+static bool isRectangularAffineForLoopNest(ArrayRef<AffineForOp> loopNest) {
+  for (AffineForOp forOp : loopNest) {
+    if (!forOp.hasConstantUpperBound() || !forOp.hasConstantLowerBound())
+      return false;
+  }
+  return true;
+}
+
+/// Returns true if this entire column of `matrix` is zero.
+static bool checkColumnIsZero(ArrayRef<SmallVector<int64_t, 4>> matrix,
+                              unsigned column) {
+  for (auto &row : matrix) {
+    if (row[column] != 0)
+      return false;
+  }
+  return true;
+}
+
+/// For a memref access function AX+B, it calculates both A and B and assigns
+/// these to protected members `loopAccessMatrices`(A) and `constVector`(B).
+/// The param `indVarIdMap` is used to locate position for coefficients of X
+/// in each row of matrix A.
+void LoopInterchange::getAffineAccessMatrices(
+    AffineForOp &rootForOp, SmallVector<Operation *, 8> &loadAndStoreOps,
+    DenseMap<Value, unsigned> &indVarIdMap, unsigned AffineForOpLoopNestSize) {
+
+  this->loopAccessMatrices.clear();
+  this->constVector.clear();
+
+  unsigned numOps = loadAndStoreOps.size();
+  for (unsigned i = 0; (i < numOps); ++i) {
```

Comment:
Why is the guard in parentheses? Drop them.

Reply:
Done.